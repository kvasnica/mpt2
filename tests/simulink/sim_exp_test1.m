function sim_exp_test1

% tests simulation of an explicit controller, regulation problems
x0 = [1; 1];
assignin('base', 'sim_x0', x0);
ref = 0;
assignin('base', 'sim_ref', ref);
Double_Integrator
assignin('base', 'sim_sysStruct', sysStruct);


%----------------------------------------------------------
% regulation problem, no deltaU constraints
clear sysStruct probStruct
Double_Integrator
probStruct.N = 3;
ctrl = mpt_control(sysStruct, probStruct);
assignin('base', 'sim_ctrl', ctrl);
sim('sim_exp');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to origin?
mbg_asserttrue(norm(sim_x(end, :)) < 1e-1);



%----------------------------------------------------------
% regulation problem, with deltaU constraints
clear sysStruct probStruct
Double_Integrator
probStruct.N = 2;
sysStruct.dumax = 0.5;
sysStruct.dumin = -0.5;
ctrl = mpt_control(sysStruct, probStruct);
assignin('base', 'sim_ctrl', ctrl);
sim('sim_exp');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to origin?
mbg_asserttrue(norm(sim_x(end, :)) < 1e-1);
% deltaU constraints satisfied?
mbg_asserttrue(max(diff(sim_u)) <= sysStruct.dumax);
mbg_asserttrue(sysStruct.dumin <= min(diff(sim_u)));


%----------------------------------------------------------
% regulation problem, no deltaU constraints, mpt_ownmpc()
clear sysStruct probStruct
Double_Integrator
probStruct.N = 3;
[C, O, V] = mpt_ownmpc(sysStruct, probStruct);
ctrl = mpt_ownmpc(sysStruct, probStruct, C, O, V);
assignin('base', 'sim_ctrl', ctrl);
sim('sim_exp');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to origin?
mbg_asserttrue(norm(sim_x(end, :)) < 1e-1);



%----------------------------------------------------------
% regulation problem, with deltaU constraints, mpt_ownmpc()
clear sysStruct probStruct
Double_Integrator
probStruct.N = 2;
sysStruct.dumax = 0.5;
sysStruct.dumin = -0.5;
[C, O, V] = mpt_ownmpc(sysStruct, probStruct);
ctrl = mpt_ownmpc(sysStruct, probStruct, C, O, V);
assignin('base', 'sim_ctrl', ctrl);
sim('sim_exp');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to origin?
mbg_asserttrue(norm(sim_x(end, :)) < 1e-1);
% deltaU constraints satisfied?
mbg_asserttrue(max(diff(sim_u)) <= sysStruct.dumax);
mbg_asserttrue(sysStruct.dumin <= min(diff(sim_u)));


