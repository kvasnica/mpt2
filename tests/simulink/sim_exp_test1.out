Adding stabilizing target set constraint...
mpt_mpqp: 19 regions
-> Generated 1 partition.

Solution consists of 19 regions

Adding stabilizing target set constraint...
Region: 20/35          
mpt_mpqp: 35 regions
-> Generated 1 partition.

Solution consists of 35 regions

Adding stabilizing target set constraint...
Solving an mpQP problem with 2 parametric, 3 optimization and 0 binary variables

mpt_mpqp: 19 regions
-> Generated 1 partition.
State vector will have the format x = [x0; u(k-1)].
Adding stabilizing target set constraint...
Solving an mpQP problem with 3 parametric, 2 optimization and 0 binary variables

Region: 20/35          
mpt_mpqp: 35 regions
-> Generated 1 partition.
